travel rule spec (#8135)
diem/diem: specifications/transactions/travel-rule/README.md

> # Off-CHain Travel Rule Exchange

Reviewer: ```suggestion
# Off-Chain Travel Rule Exchange
```

> This protocol is used to exchange KYC information about a potential transaction between to parties that will be resolved on-chain upon its completion. This is a dual attestation, preflight protocol. At the end of the protocol, the sender will receive a signature from the recipient so that the Diem Paymnet Network can verify that the tranasfer has been signed off by both the sender and recipient. Some considerations:

Reviewer: ```suggestion
This protocol is used to exchange KYC information about a potential transaction between to parties that will be resolved on-chain upon its completion. This is a dual attestation, preflight protocol. At the end of the protocol, the sender will receive a signature from the recipient so that the Diem Payment Network can verify that the transfer has been signed off by both the sender and recipient. Some considerations:
```

> * Participants that do not implement this cannot tranfser amounts above the travel rule limit

Reviewer: ```suggestion
* Participants that do not implement this cannot transfer amounts above the travel rule limit
```

Reviewer: This seems a bit harsh no?  Isn't it rather than this is a potential protocol for transferring amounts above the travel rule limit?

Reviewer 2: @aching If we aren't going to make this *the* standard for transacting on diem then what's the point? Just let every VASP figure it out themselves and we'll go and work on other things.

Reviewer 3: above => equal or above

Reviewer: I agree we want people to opt-in to this protocol (as no other option currently exists).  However, it is not required for DPN members specifically AFAIK.  The summary of DIP-1 indicates this to me as well - https://github.com/diem/dip/blob/master/dips/dip-1.mdx#summary

Author: I think there are three things:
* This is the only standard supported
* One must participate in some dual attestation protocol
* We need another doc that actually covers transaction flows, since this is not a transaction flow

* Participants may reject travel rule requests for transfers below the travel rule limit

## PaymentCommand

> All requests between participants are structured as `PaymentCommand`s within `CommandRequestObject`s:

Reviewer: ```suggestion
All requests between participants are structured as `PaymentCommand`s within `CommandRequestObject`s (as shown in the example below):
```

```
{
    "_ObjectType": "CommandRequestObject",
    "command_type": "PaymentCommand",
    "cid": "88b282d6-1811-29f6-82be-0421d0ee9887",
    "command": {
	    "_ObjectType": "PaymentCommand",
	    "payment": {
		    "sender": {
			    "address": "dm1pg9q5zs2pg9q5zs2pg9q5zs2pg9skzctpv9skzcg9kmwta",
			    "kyc_data": {
				    "payload_version": 1,
				    "type": "individual",
				    "given_name": "ben",
				    "surname": "maurer",
				    "address": {
					"city": "Sunnyvale",
					"country": "US",
					"line1": "1234 Maple Street",
					"line2": "Apartment 123",
					"postal_code": "12345",
					"state": "California",
				    },
				    "dob": "1920-03-20",
				    "place_of_birth": {
					"city": "Sunnyvale",
					"country": "US",
					"postal_code": "12345",
					"state": "California",
				    }
				},
			    	"status": {
			    		"status": "ready_for_settlement",
			    	}
			},
		    "receiver": {
			    "address": "dm1pgfpyysjzgfpyysjzgfpyysjzgf3xycnzvf3xycsm957ne",

> 			},

Reviewer 3: may reformat above for reader

		    "action": {
			    "amount": 100,
			    "currency": "USD",
			    "action": "charge",
			    "timestamp": 72322,
			},
		    "description": "A free form or structured description of the payment.",
		},
	},
}
```

A successful response to this command should look like the following:

```
{
    "_ObjectType": "CommandResponseObject",
    "status": "success",
}
```

A travel-rule exchange may require multiple `PaymentCommand`'s in order to complete. In these instances, the underlying structures may be passed back and forth many times. As outlined in the rest of this section, certain fields must be defined at the outset and others can be defined later. Some fields may be adjusted or mutated throughout the exchange, whereas others must remain immutable.

### PaymentCommand Object

| Field          | Type            | Required | Description |
|-------         |------           |----------|-------------|
| `_ObjectType`  | `str`           | Y        | The fixed string `PaymentCommand` |
| `payment`      | `PaymentObject` | Y        | Contains a `PaymentObject` |

```
{
    "_ObjectType": "PaymentCommand",
    "payment": {
        PaymentObject(),
    }
}
```

### PaymentObject

> Some fields are immutable after they are defined once. Others can be updated multiple times (see below). Updating immutable fields with a different value results in a Command error.

Reviewer: I do like the column simplicity.  Do you think it would be helpful to have an (Im)mutable column in the table below to make it more clear about which are immutable or not?

| Field                           | Type                  | Required | Description |
|-------                          |------                 |----------|-------------|

> | `sender`/`receiver`             | `PaymentActorObject`  | Y        | Information about the sender/receiver in this payment |

Reviewer: Something that might also be helpful is to consider adding links to the objects to make it easier for the reader to find the types.

Author: next revision

> | `sender`/`receiver`             | `PaymentActorObject`  | Y        | Information about the sender/receiver in this payment |

Reviewer 3: "sender/receiver" should be 2 fields, merged into one line is like one field.

Author: hahaha, copy paste from the dip, but I agree and wondered if there was some reason for the single row

|`reference_id`                   | `str`                 | Y        | Unique reference ID of this payment on the payment initiator VASP (the VASP which originally created this payment Object). This value should be globally unique. This field is mandatory on payment creation and immutable after that. This must be a UUID according to RFC4122 with "-"s included. |
| `original_payment_reference_id` | `str`                 | N        | Used to refer an old payment known to the other VASP. For example, used for refunds. The reference ID of the original payment will be placed into this field. This field is mandatory on refund and immutable |

> | `recipient_signature`           | `str`                 | N        | Signature of the recipient of this transaction encoded in hex. The is signed with the compliance key of the recipient VASP and is used for on-chain attestation from the recipient party. This may be omitted on blockchains which do not require on-chain attestation. Generated via Recipient Signature |

Reviewer: Is there more details on what exactly is being signed by the recipient?

Reviewer 2: This is a good point, we really should probably either include the spec of what is being signed or provide a link to it here.

Author: yup, we need a guide / spec that connects all these things together

    "line2": "Apartment 123",
    "postal_code": "12345",
    "state": "California",
}
```

### NationalIdObject

Represents a national ID.

| Field      | Type  | Required | Description |
|-------     |------ |----------|-------------|
| `id_value` | `str` | Y        | Indicates the national ID value,for example, a social security number |
| `country`  | `str` | N        | [Two-letter country code](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) |
| `type`     | `str `| N        | Indicates the type of the ID |

```
{
    "id_value": "123-45-6789",
    "country": "US",
    "type": "SSN",
}
```

### PaymentActionObject

| Field       | Type   | Required | Description |
|-------      |------  |----------|-------------|
| `amount`    | `uint` | Y        | Amount of the transfer as defined by the on-chain base units. For example, if XUS is represented on-chain as "1" equals 1e-6 units, then "1" here would also equal 1e-6 units. |
| `currency`  | `enum` | Y        | One of the supported currency types, for example, XUS |
| `action`    | `enum` | Y        | Populated in the request. This value indicates the requested action to perform, and the only valid value is charge. |

> | `timestamp` | `uint` | Y        | Unix time indicating the time that the payment `Command` was created. |

Reviewer: This wasn't obvious to me how timestamp is used by the sender/receiver.  Just for record keeping?

Author: @xli ?

```
{
    "amount": 100,
    "currency": "USD",
    "action": "charge",
    "timestamp": 72322,
}
```

### StatusObject

| Field           | Type   | Required | Description |
|-------          |------  |----------|-------------|
| `status`        | `enum` | Y        | Status of the payment from the perspective of this actor. This field can only be set by the respective sender/receiver VASP and represents the status on the sender/receiver VASP side. This field is mandatory by this respective actor (either sender or receiver side) and mutable. Valid values are specified in `StatusEnum` |
| `abort_code`    | `enum` | N        | In the case of an `abort` status, this field may be used to describe the reason for the abort. Represents the error code of the corresponding error. Valid values are specified in `AbortCodeEnum` |
| `abort_message` | `str`  | N        | Additional details about this error. To be used only when `code` is populated |

```
{
    "status": "needs_kyc_data",
}
```

#### AbortCodeEnum

| Error code | Description |
|------------|-------------|

> | `rejected` | The payment is rejected. Note, it must must not be used in the `original_payment_reference_id` field of future payments. |

Reviewer: ```suggestion
| `rejected` | The payment is rejected. Note, it must not be used in the `original_payment_reference_id` field of future payments. |
```

Is one must enough? =)

Author: leave no ambiguity

| `soft_match`           | Actor’s KYC data resulted in a soft-match. It is suggested that soft matches are resolved within 24 hours. |

### Travel Rule Errors

The Travel Rule Exchange extends the Command Errors defined in the off-chain protocol with the following:

| Error code                              | Description |
|------------                             |-------------|
| `invalid_command_producer`              | The HTTP request sender is not the right actor to send the `PaymentObject`. For example, if the receiver sends a new command with `PaymentObject` change that should be done by the sender. |
| `invalid_initial_or_prior_not_found`    | Could not find command by `reference_id` for a non-initial state command object; for example, the receiver received a `PaymentCommand` object that the sender status is `ready_for_settlement`, but the receiver is unable to find a command for the given `reference_id`. |
| `no_kyc_needed`                         | Payment action amount is under travel rule limit. |
| `unknown_address`                       | Unknown `sender.address` or `receiver.address`. |

> | `conflict`                              | Either the a `cid` has been reused or data is unavailable due to competing requests. |

Reviewer 3: `the a`

* `sender.status.status` = `need_kyc_data`.
* `sender.kyc_data` = The `KycDataObject` representing the sender.
* `receiver.status.status` = `none`

**SINIT -> RSEND (Step 4)**

The Receiver issues `PaymentCommand` to update an existing payment.

The receiver VASP examines the `sender.kyc_data` object, and is satisfied that given the sender information the payment can proceed. It responds with a Payment Command that includes:

* `receiver.status.status` = `ready_for_settlement`
* `receiver.kyc_data` = The `KycDataObject` representing the receiver.
* `recipient_signature` = a valid recipient signature.

**SINIT -> RABORT (Step 9)**

The Receiver issues `PaymentCommand` to update an existing payment.

The receiver VASP examines the sender KYC information and is either not satisfied the payment can proceed, needs more time to process the KYC information, or requires additional information to determine if the payment can proceed. It includes a command to abort the payment with an appropriate error code.

* `receiver.status.status`: `abort`
* `receiver.status.abort_code`: one of `no-kyc-needed` or `rejected`.

> The sender can initiate a payment on-chain in case of an abort with `no-kyc-needed`.

Reviewer: Is it considered a bug from the sender side that the abort_code is no-kyc-needed?

Shouldn't the sender always know whether kyc is needed or does it depend on the receiver?

Author: great question, I think there are a lot of edge cases that the current frameworks don't resolve, namely, what happens if I request on an amount below TR or how to handle TR requirements for different regions. There's a lot of work that needs to be done and I suspect we'll likely find ourselves in a situation where we need a lightweight pre-flight before any transaction just to determine if we actually need KYC.
